src/common/messageBundles.js

```diff
@@ -0,0 +1,240 @@
+ *     en: {
+ *       "gpii_psp_header_autosaveText": "Auto-save is on",
+ *       ...
+ *     }
+ * It means that the "Auto-save is on" message should be the value of the `autosaveText`
+ * property within the model's messages object of the "gpii.psp.header" component.
+ *
+ * This component has a `messageDistributor` subcomponent which is created
+ * programmatically and it takes care of providing the necessary messages to the
+ * components which need them via dynamically generated `distributeOptions` blocks.
+ * Note that the messages that a component uses must to be located in the "model".
+ * This approach makes use of the modelListeners in order for re-rendering to happen
+ * when the locale changes.
+ */
+fluid.defaults("gpii.app.messageBundles", {
+    gradeNames: ["fluid.modelComponent", "{that}.options.messageDistributorGrade"],
+
+    model: {
+        locale: null, // the defaultLocale will be used initially
+        messages: {}
+    },
+
+    defaultLocale: "en",
+
+    messageBundlesPath: "build/gpii-app-messageBundles.json",
+
+    messageBundles: "@expand:gpii.app.messageBundles.loadMessageBundles({that}.options.messageBundlesPath)",
+
+    modelListeners: {
+        locale: {
+            func: "{that}.updateMessages"
+        }
+    },
+
+    messageDistributorGrade: {
+        expander: {
+            funcName: "gpii.app.messageBundles.getMessageDistributorGrade",
+            args: [
+                "{that}.options.messageBundles",
+                "{that}.options.defaultLocale"
+            ]
+        }
+    },
+
+    invokers: {
+        updateMessages: {
+            funcName: "gpii.app.messageBundles.updateMessages",
+            args: [
+                "{that}",
+                "{that}.options.messageBundles",
+                "{that}.model.locale",
+                "{that}.options.defaultLocale"
+            ]
+        }
+    }
+});
+
+/**
+ * This function creates a `messageDistributor` grade which has a dynamically generated
+ * `distributeOptions` block. The name of the grade is returned as a result of this
+ * function and is applied to the `messageBundles` component. By using this raw dynamic grade
+ * mechanism this component is constructed before the other components. This is important
+ * because distributeOptions will be considered only at creation time. This means that the
+ * distributor should be created before any other components to which it distributes
+ * messages.
+ * @param messageBundles {Object} An object containing all i18n messages for all supported
+ * locales
+ * @param defaultLocale {String} A string representing the default locale.
+ * @return {String} The name of the distributor grade.
+ */
+gpii.app.messageBundles.getMessageDistributorGrade = function (messageBundles, defaultLocale) {
+    var defaultMessages = messageBundles[defaultLocale],
+        distributions = gpii.app.messageBundles.getMessageDistributions(defaultMessages);
+
+    fluid.defaults("gpii.app.messageDistributor", {
```

Comment:
This is pretty interesting. On one hand, would prefer this not to be an "app-wide singleton grade" since the scope of an Infusion application might be quite large - what we do in other cases is append a guid to this name to ensure that it can't collide. On the other hand this creates an issue with respect to "garbage collecting" this grade once it is not referenced any more, so some more thought is required here - I'll create a JIRA and link here. In the meantime I think the approach with a guid is preferable.

Reply:
Can you add a comment linking this to https://issues.fluidproject.org/browse/FLUID-6273 as a place that will be reworked someday and instead make the defaults name as "gpii.app.messageDistributor." + fluid.allocateGuid() - thanks